Update parallelism.md (#13892)
huggingface/transformers: docs/source/parallelism.md

It performs a sort of 4D Parallelism over Sample-Operator-Attribute-Parameter.

1. Sample = Data Parallelism (sample-wise parallel)

> 2. Operator = Parallelize one operation into several tasks

Reviewer: ```suggestion
2. Operator = Parallelize a single operation into several sub-operations
```

3. Attribute = Data Parallelism (length-wise parallel)

> 4. Parameter = Model Parallelism (it doesn't matter dimension like horizontal or vertical)

Reviewer: ```suggestion
4. Parameter = Model Parallelism (regardless of dimension - horizontal or vertical)
```

> ### Examples

Reviewer: ```suggestion
Examples:
```

* Sample

> We have 10 batches of 512 length. If we parallelize them by sample dimension into 2 devices, 10 x 512 will be 5 x 2 x 512.

Reviewer: ```suggestion
Let's take 10 batches of sequence length 512. If we parallelize them by sample dimension into 2 devices, we get 10 x 512 which becomes be 5 x 2 x 512.
```

* Operator

> If we perform layer normalization, we compute std first and mean second, finally we can normalize data. At this time, computing std and mean can be parallelize. So if we parallelize them by operator dimension into 2 devices (cuda:0, cuda:1), firstly we copied input data into both devices, and cuda:0 computes std, cuda:1 computes mean at the same time.

Reviewer: ```suggestion
If we perform layer normalization, we compute std first and mean second, and then we can normalize data. Operator parallelism allows computing std and mean in parallel. So if we parallelize them by operator dimension into 2 devices (cuda:0, cuda:1), first we copy input data into both devices, and cuda:0 computes std, cuda:1 computes mean at the same time.
```

* Attribute

We have 10 batches of 512 length. If we parallelize them by attribute dimension into 2 devices, 10 x 512 will be 10 x 2 x 256.

* Parameter

> It is similar with tensor model parallelism or niave layer-wise model parallelism.

Reviewer: ```suggestion
It is similar with tensor model parallelism or naive layer-wise model parallelism.
```
